llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -200,93 +171,142 @@ groupEntryPointsByKernelType(ModuleDesc &MD,
   return EntryPointGroups;
 }
 
-// Represents a call graph between functions in a module. Nodes are functions,
-// edges are "calls" relation.
-class CallGraph {
+// Represents "dependency" or "use" graph of global objects (functions and
+// global variables) in a module. It is used during device code split to
+// understand which global variables and functions (other than entry points)
+// should be included into a split module.
+//
+// Nodes of the graph represent LLVM's GlobalObjects, edges "A" -> "B" represent
+// the fact that if "A" is include into a module, then "B" should be included
+// as well.
+//
+// Examples of dependencies which are represented in this graph:
+// - Function FA calls function FB
+// - Function FA uses global variable GA
+// - Global variable GA references (intiialized with) function FB
+// - Function FA stores address of a function FB somewhere
+// - Function FA uses global variable GB
+//
+// The following cases are treated as dependencies between global objects:
+// 1. Global object A is used within by a global object B in any way (store,
+//    bitcast, phi node, call, etc.): "A" -> "B" edge will be added to the
+//    graph;
+// 2. function A performs an indirect call of a function with signature S and
+//    there is a function B with signature S. "A" -> "B" edge will be added to
+//    the graph;
+class DependencyGraph {
 public:
-  using FunctionSet = SmallPtrSet<const Function *, 16>;
-
-private:
-  std::unordered_map<const Function *, FunctionSet> Graph;
-  SmallPtrSet<const Function *, 1> EmptySet;
-  FunctionSet AddrTakenFunctions;
+  using GlobalSet = SmallPtrSet<const GlobalValue *, 16>;
+
+  DependencyGraph(const Module &M) {
+    // Group functions by their signature to handle case (2)
+    DenseMap<const FunctionType *, DependencyGraph::GlobalSet>
+        FuncTypeToFuncMap;
+    for (const auto &F : M.functions()) {
+      FuncTypeToFuncMap[F.getFunctionType()].insert(&F);
+    }
 
-public:
-  CallGraph(const Module &M) {
-    for (const auto &F : M) {
-      for (const Value *U : F.users()) {
-        if (const auto *I = dyn_cast<CallInst>(U)) {
-          if (I->getCalledFunction() == &F) {
-            const Function *F1 = I->getFunction();
-            Graph[F1].insert(&F);
-          }
-        }
-      }
-      if (F.hasAddressTaken()) {
-        AddrTakenFunctions.insert(&F);
+    // We add every function into the graph
+    for (const auto &F : M.functions()) {
+      // case (1)
+      for (const Value *U : F.users())
+        addUserToGraphRecursively(cast<const User>(U), &F);
+
+      // case (2)
+      for (const auto &I : instructions(F)) {
+        if (!isa<CallInst>(&I))
+          continue;
+
+        const auto *CI = cast<CallInst>(&I);
+        // Direct calls were handled above
+        if (!CI->isIndirectCall())
+          continue;
+
+        const FunctionType *Signature = CI->getFunctionType();
+        const auto &PotentialCallees = FuncTypeToFuncMap[Signature];
```

Comment:
Good question. If I understand correctly, `Type`s are immutable in LLVM within `LLVMContext` and only created ones for each unique LLVM IR data type. From [doxygen](https://llvm.org/doxygen/classllvm_1_1Type.html#details):

> The instances of the [Type](https://llvm.org/doxygen/classllvm_1_1Type.html) class are immutable: once they are created, they are never changed.

>
> Also note that only one instance of a particular type is ever created. Thus seeing if two types are equal is a matter of doing a trivial pointer comparison. To enforce that no two equal instances are created, [Type](https://llvm.org/doxygen/classllvm_1_1Type.html) instances can only be created via static factory methods in class [Type](https://llvm.org/doxygen/classllvm_1_1Type.html) and in derived classes. Once allocated, Types are never free'd.

So, the question is: is it legal to call a function through a function pointer of a different type? [expr.reinterpret.cast.6](https://eel.is/c++draft/expr.reinterpret.cast#6) says that it is a UB:

> A function pointer can be explicitly converted to a function pointer of a different type[.](https://eel.is/c++draft/expr.reinterpret.cast#6.sentence-1)
> [Note [4](https://eel.is/c++draft/expr.reinterpret.cast#note-4): The effect of calling a function through a pointer to a function type ([[dcl.fct]](https://eel.is/c++draft/dcl.fct)) that is not the same as the type used in the definition of the function is undefined ([[expr.call]](https://eel.is/c++draft/expr.call))[.](https://eel.is/c++draft/expr.reinterpret.cast#6.sentence-2) — end note]

When virtual functions are used, it is allowed to use _co-variant_ return types, from [class.virtual.8](https://eel.is/c++draft/class.virtual#8): 

> (http://github.com/Eelis/draft/tree/fb4dab8b7ad7b6df0790ba9a808d38c312bc8fc7/source/classes.tex#L3858)
The return type of an overriding function shall be either identical to the return type of the overridden function or [covariant](https://eel.is/c++draft/class.virtual#def:return_type,covariant) with the classes of the functions[.](https://eel.is/c++draft/class.virtual#8.sentence-1) If a function D​::​f overrides a function B​::​f, the return types of the functions are covariant if they satisfy the following criteria:
>- both are pointers to classes, both are lvalue references to classes, or both are rvalue references to classes [102](https://eel.is/c++draft/class.virtual#footnote-102)
> - the class in the return type of B​::​f is the same class as the class in the return type of D​::​f, or is an unambiguous and accessible direct or indirect base class of the class in the return type of D​::​f
> - both pointers or references have the same cv-qualification and the class type in the return type of D​::​f has the same cv-qualification as or less cv-qualification than the class type in the return type of B​::​f[.](https://eel.is/c++draft/class.virtual#8.sentence-2)

This concerns me a bit, because it means that we potentially may have different signatures (at LLVM IR level) for virtual functions overloads. They should be the same if opaque pointers are enabled, but we are not there yet. I will do some experiments to see if that case is indeed a problem here

Reply:
cool, thanks

Reply:
Though about this more, I don't think that the issue with virtual function can actually be reproduced, added comments clarifying that in e2aecc9d976610ab1b97e82a6449b473241f5eae

Reply:
I've just realized that global variable which holds `vtable` is only directly referenced during polymorphic object construction, but not during calls to virtual member functions. Let's say that you construct a polymorphic object in kernel `A`, store it in memory and then perform a call to a virtual member function `foo` of it in kernel `B`. We would draw a link between kernel `A` and virtual function `foo`, because `A` uses `vtable`, which references `foo`. However, kernel `B` won't directly reference `vtable`, it will simply load it from the object itself and our only chance to draw a link between kernel `B` and virtual function `foo` is using signatures matching. If covariant types are involved, this won't work.

Unfortunately, I don't have a solution for covariant return types right now and I'm not sure if it worth investing resources into, because we expect to switch to opaque pointers soon. The problem with covariant return types doesn't exists with opaque pointers.

Reply:
It looks like one must consider variance in the parameter types as well: https://godbolt.org/z/Wf7oW9o81

in `call_foo`, there is the instruction

```
 %4 = load %struct.base* (%struct.base*)*, %struct.base* (%struct.base*)** %3, align 8, !dbg !29
 %5 = tail call noundef %struct.base* %4(%struct.base* noundef nonnull align 8 dereferenceable(8) %0), !dbg !29
```

so `call_foo` can potentially call functions with type ` %struct.base* (%struct.base*)`. This misses the override `%struct.derived* @_ZN7derived3fooEv(%struct.derived* ...)`

Reply:
@jzc, thanks, good catch. The first argument is essentially a `this` pointer and apparently we pass a pointer to base here, since we don't know which exact overload we are calling.

This is more concerning, because it would effectively prevent proper bundling of virtual functions, because of broken signatures matching.

Reply:
To share my latest thoughts on this: we don't have a guarantee that a call through function pointer will work, if caller and callee are within different device images. This restriction will be in one or another form included into the language specification for virtual functions and there it would mean that kernel which allocates an object with vtable must be in the same device image as kernel which uses virtual functions from that object.

Since virtual functions are traceable from kernel which does the allocation and it has to be together with kernel which does the call, both these issues (with co-variant return types and type of `this` pointer) won't cause any issues even when opaque pointers are disables.